lib/spack/spack/schema/packages.py

```diff
@@ -168,6 +168,7 @@
                 # attribute, it could be useful to set it here
                 "package_attributes": package_attributes,
                 "variants": variants,
+                "allow_compiler_mixing": {"type": "boolean", "default": False},
```

Comment:
Is this intended to support packages that fail to build with certain compilers? I think that might not be sufficiently flexible for things that can build with some compilers (and should have mixing enforced for those compilers) but not others.

I wonder if this case would be better supported by a "strong preference" mode for `concretizer:compiler_mixing` instead of an exception to the rule.

Reply:
I think the simplest implementation would be to have a high priority optimization criteria for minimizing the count. You would compute it the same way as here, but using a `#min` optimization statement.

It would probably require something slightly fancy to ensure we treat 0 and 1 as equally good, so that the solver doesn't turn off options to avoid using a compiler at all.

Reply:
This configuration seems confusing to me:
```yaml
concretizer:
    compiler_mixing: false
packages:
    hdf5:
        allow_compiler_mixing: true
```
for the following reasons:
1. The configuration in `concretizer` seems to be contradicted by the one in `packages`
2. We need to inspect 2 sections to understand what is the behavior with compiler mixing (I would keep everything within the `concretizer` section)
3. It is not clear what would be the behavior with respect to dependencies and dependents of `hdf5`. Do they need to use the same compiler as `hdf5`? How is the compiler for `hdf5` selected?

Reply:
I'm happy to go with above proposal or, to address

> The downside is the exceptions block is a no-op when allow: true

the schema could be

```
concretizer:
  # if this is true, everything can mix
  # if this is false, nothing can mix
  # if this is a list of strings, only the packages mentioned in the list can be mixed
  compiler_mixing: [bool, list] 
```